url542ae48576_tonyheadford_m68k_tgz-pJ8-m68k_MonitorJ8/src/m68k/cpu/instructions/BCHG.java

```diff
@@ -0,0 +1,276 @@
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+*/
+
+public class BCHG implements InstructionHandler
+{
+	protected final Cpu cpu;
+
+	public BCHG(Cpu cpu)
+	{
+		this.cpu = cpu;
+	}
+
+	public void register(InstructionSet is)
+	{
+		int base = 0x0140;
+		Instruction i;
+
+		// dynamic mode
+		for(int ea_mode = 0; ea_mode < 8; ea_mode++)
+		{
+			if(ea_mode == 1)
+				continue;
+
+			if(ea_mode == 0)
+			{
+				// data register destination
+				i = new Instruction() {
+					public int execute(int opcode)
+					{
+						return bchg_dyn_long(opcode);
+					}
+					public DisassembledInstruction disassemble(int address, int opcode)
+					{
+						return disassembleOp(address, opcode, Size.Long);
+					}
+				};
+			}
+			else
+			{
+				i = new Instruction() {
+					public int execute(int opcode)
+					{
+						return bchg_dyn_byte(opcode);
+					}
+					public DisassembledInstruction disassemble(int address, int opcode)
+					{
+						return disassembleOp(address, opcode, Size.Byte);
+					}
+				};
+			}
+			for(int ea_reg = 0; ea_reg < 8; ea_reg++)
+			{
+				if(ea_mode == 7 && ea_reg > 1)
+					break;
+
+				for(int r = 0; r < 8; r++)
+				{
+					is.addInstruction(base + (r << 9) + (ea_mode << 3) + ea_reg, i);
+				}
+
+			}
+		}
+
+		// static mode
+		base = 0x0840;
+		for(int ea_mode = 0; ea_mode < 8; ea_mode++)
+		{
+			if(ea_mode == 1)
+				continue;
+
+			if(ea_mode == 0)
+			{
+				// data register destination
+				i = new Instruction() {
+					public int execute(int opcode)
+					{
+						return bchg_static_long(opcode);
+					}
+					public DisassembledInstruction disassemble(int address, int opcode)
+					{
+						return disassembleOp(address, opcode, Size.Long);
+					}
+				};
+			}
+			else
+			{
+				i = new Instruction() {
+					public int execute(int opcode)
+					{
+						return bchg_static_byte(opcode);
+					}
+					public DisassembledInstruction disassemble(int address, int opcode)
+					{
+						return disassembleOp(address, opcode, Size.Byte);
+					}
+				};
+			}
+			for(int ea_reg = 0; ea_reg < 8; ea_reg++)
+			{
+				if(ea_mode == 7 && ea_reg > 1)
+					break;
+
+				is.addInstruction(base + (ea_mode << 3) + ea_reg, i);
+			}
+		}
+
+	}
+
+	protected final int bchg_dyn_byte(int opcode)
+	{
+		// for memory destination, the bitnbr is MOD 8 - for data reg destination, the bitnbr is mod 32
+		int bit = cpu.getDataRegisterLong((opcode >> 9) & 0x07) & 7;
+		bit = 1 << bit;
+		// memory destination
+		Operand op = cpu.resolveDstEA((opcode >> 3) & 0x07, (opcode & 0x07), Size.Byte);
+		int val = op.getByte();
+
+		// Z_FLAG set according to original value
+		if((val & bit) != 0)
+		{
+			cpu.clrFlags(Cpu.Z_FLAG);
+			//clear the bit
+			val &= ~(bit);
+		}
+		else
+		{
+			cpu.setFlags(Cpu.Z_FLAG);
+			//set the bit
+			val |= bit;
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 145:156, 172:183, 201:212, 228:239.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.